### PR TITLE
AlertingNG: Slack notification channel

### PR DESCRIPTION
**Pull request by codesome**
**What this PR does / why we need it**:

Adds slack notification channel. Things are taken from both Prometheus AM implementation and Grafana implementation.

### CODE REVIEW

**Comment by codesome** [comment]
Once https://github.com/grafana/grafana/pull/32604/ is merged, I will rebase this PR to plug slack into the notification pipeline

File: pkg/services/ngalert/notifier/channels/slack.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+		IconURL:   tmpl(sn.IconURL),
+		Attachments: []attachment{
+			{
+				Color:      getAlertStatusColor(alerts.Status()),
+				Title:      tmpl(sn.Title),
+				Fallback:   tmpl(sn.Fallback),
+				Footer:     "Grafana v" + setting.BuildVersion,
+				FooterIcon: "https://grafana.com/assets/img/fav32.png",
+				Ts:         time.Now().Unix(),
+				TitleLink:  "TODO: rule URL",
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/channels/slack.go:193]: Can we leave these fields empty with a TODO on top explaining what's suppose to go here and at which point do we expect it to have it available. 

Even better if this is something that must come from the evaluation engine, for example, let's create an issue so we can assign it.

File: pkg/services/ngalert/notifier/channels/slack.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+				Text:       tmpl(sn.Text),
+				Fields:     nil, // TODO. Should be a config.
+			},
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/channels/slack.go:196]: Ditto from above.

File: pkg/services/ngalert/notifier/channels/slack.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	Fields     []config.SlackField `json:"fields,omitempty"`
+	Footer     string              `json:"footer"`
+	FooterIcon string              `json:"footer_icon"`
+	Color      string              `json:"color,omitempty"`
+	Ts         int64               `json:"ts,omitempty"`
+}
+
+func getAlertStatusColor(status model.AlertStatus) string {
+	if status == model.AlertFiring {
+		return "#D63232"
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/channels/slack.go:137]: Should these be configurable?

Comment (codesome) [reply pkg/services/ngalert/notifier/channels/slack.go:137]: Currently we used fixed colors, so I think it need not be configurable even now

Comment (gotjosh) [reply pkg/services/ngalert/notifier/channels/slack.go:137]: Let's move them to a constant then 😄

File: pkg/services/ngalert/notifier/channels/slack.go (hunk 1 above)
Comment (codesome) [reply pkg/services/ngalert/notifier/channels/slack.go:191]: Not 100% sure what goes on here, so I will create an issue for all the TODOs

File: pkg/services/ngalert/notifier/channels/slack.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	Fields     []config.SlackField `json:"fields,omitempty"`
+	Footer     string              `json:"footer"`
+	FooterIcon string              `json:"footer_icon"`
+	Color      string              `json:"color,omitempty"`
+	Ts         int64               `json:"ts,omitempty"`
+}
+
+const (
+	ColorAlertFiring = "#D63232"
+	ColorAlertResolved
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/channels/slack.go:135]: This does not have a color is it intended?

File: pkg/services/ngalert/notifier/channels/slack.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+		Username:  tmpl(sn.Username),
+		IconEmoji: tmpl(sn.IconEmoji),
+		IconURL:   tmpl(sn.IconURL),
+		Attachments: []attachment{
+			{
+				Color:      getAlertStatusColor(alerts.Status()),
+				Title:      tmpl(sn.Title),
+				Fallback:   tmpl(sn.Fallback),
+				Footer:     "Grafana v" + setting.BuildVersion,
+				FooterIcon: "https://grafana.com/assets/img/fav32.png",
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/channels/slack.go:194]: should this also be a constant?

**Review by gotjosh — APPROVED** [review]

File: pkg/services/ngalert/notifier/channels/slack.go (hunk 4 above)
Comment (codesome) [reply pkg/services/ngalert/notifier/channels/slack.go:135]: My bad, it does have a color

File: pkg/services/ngalert/notifier/channels/slack.go (hunk 5 above)
Comment (codesome) [reply pkg/services/ngalert/notifier/channels/slack.go:194]: I think this should be fine here. If I find it being used in multiple places, I will make it a constant.

Comment (codesome) [reply pkg/services/ngalert/notifier/channels/slack.go:194]: Or let me just make it a constant :)

**Pull request merged**